Support for CUDA emulator for AMD GPUs

AznamirWoW opened the pull request:
disabled CUDNN when Zluda is detected
added fallback for FFT since it is not supported by HIP SDK
preventing HIP Crash by jit decorator

blaisewf commented:
@deiteris is an amd pro, maybe he can share some thoughts about this changes, because i can’t test them 😅

blaisewf commented:
also, the installer needs some modifications to install all the necessary libraries

deiteris commented on rvc/lib/predictors/RMVPE.py:
```diff
@@ -408,15 +409,30 @@ def forward(self, audio, keyshift=0, speed=1, center=True):
-        )
+        # Zluda fall-back to CPU for FFTs since HIP SDK has no cuFFT alternative
+        if audio.device.type == "cuda" and torch.cuda.get_device_name(audio.device.index).endswith("[ZLUDA]"):
+            fft = torch.stft(
```
Simply move `hann_window` and `audio` to CPU before using torch.stft. Like:
```py
source_device = audio.device
if audio.device.type == "cuda" and torch.cuda.get_device_name(audio.device.index).endswith("[ZLUDA]"):
  audio = audio.to('cpu')
  self.hann_window[keyshift_key] = self.hann_window[keyshift_key].to('cpu')
fft = torch.stft(
    audio,
    n_fft=n_fft_new,
    hop_length=hop_length_new,
    win_length=win_length_new,
    window=self.hann_window[keyshift_key],
    center=center,
    return_complex=True,
).to(source_device) # no-op if same device
```

Or use backend-agnostic implementation of STFT.
https://github.com/RVC-Project/Retrieval-based-Voice-Conversion-WebUI/blob/main/infer/lib/rmvpe.py#L29-L156

AznamirWoW replied:
> Simply move `hann_window` and `audio` to CPU before using torch.stft. Like:
> 
> ```python
> source_device = audio.device
> if audio.device.type == "cuda" and torch.cuda.get_device_name(audio.device.index).endswith("[ZLUDA]"):
>   audio = audio.to('cpu')
>   self.hann_window[keyshift_key] = self.hann_window[keyshift_key].to('cpu')
> fft = torch.stft(
>     audio,
>     n_fft=n_fft_new,
>     hop_length=hop_length_new,
>     win_length=win_length_new,
>     window=self.hann_window[keyshift_key],
>     center=center,
>     return_complex=True,
> ).to(source_device) # no-op if same device
> ```
> 
> Or use backend-agnostic implementation of STFT. https://github.com/RVC-Project/Retrieval-based-Voice-Conversion-WebUI/blob/main/infer/lib/rmvpe.py#L29-L156

this change results in

  File "G:\GitNew\Applio-Zluda\rvc\train\mel_processing.py", line 85, in spectrogram_torch
    spec = torch.stft(
  File "G:\GitNew\Applio-Zluda\env\lib\site-packages\torch\functional.py", line 650, in stft
    return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
TypeError: stft() received an invalid combination of arguments - got (tuple, int, int, int, Tensor, bool, bool, bool), but expected one of:
 * (Tensor input, int n_fft, int hop_length, int win_length, Tensor window, bool center, str pad_mode, bool normalized, bool onesided, bool return_complex)
 * (Tensor input, int n_fft, int hop_length, int win_length, Tensor window, bool normalized, bool onesided, bool return_complex)
      didn't match because some of the arguments have invalid types: (tuple of (Tensor,), int, int, int, Tensor, bool, bool, bool)

deiteris replied:
`audio` became a tuple of Tensor? 🤔 Looks strange, how exactly did you apply the change?

AznamirWoW replied:
> `audio` became a tuple of Tensor? 🤔 Looks strange, how exactly did you apply the change?

sorry, made a typo appying the new changes, all good

deiteris commented on rvc/train/extract/extract_feature_print.py:
```diff
@@ -52,6 +52,14 @@ def read_wave(wav_path, normalize=False):
 model = models[0]
 model = model.to(device)
+
+if torch.cuda.is_available() and torch.cuda.get_device_name("cuda:0").endswith("[ZLUDA]"):
```
```suggestion
if torch.cuda.is_available() and torch.cuda.get_device_name(device).endswith("[ZLUDA]"):
```

deiteris commented on rvc/train/mel_processing.py:
```diff
@@ -76,19 +77,35 @@ def spectrogram_torch(y, n_fft, hop_size, win_size, center=False):
-    )
-
+    # Zluda fall-back to CPU for FFTs since HIP SDK has no cuFFT alternative
+    if y.device.type == "cuda" and torch.cuda.get_device_name(y.device.index).endswith("[ZLUDA]"):
```
Same as in [rvc/lib/predictors/RMVPE.py](https://github.com/IAHispano/Applio/pull/513/files#diff-de128a8d02580f48332bab47d72e4d9af0892ec40cb2d7148d9ad05f55d820ee)

deiteris commented on rvc/train/train.py:
```diff
@@ -236,6 +236,13 @@ def run(
     if torch.cuda.is_available():
         torch.cuda.set_device(rank)
 
+    if torch.cuda.is_available() and torch.cuda.get_device_name("cuda:0").endswith("[ZLUDA]"):
```
1. Merge with condition above.
2. Since current device is set, simply invoke `torch.cuda.get_device_name()`. It will get the current set device:
    ```suggestion
        if torch.cuda.get_device_name().endswith("[ZLUDA]"):
    ```

deiteris commented:
Unfortunately I cannot test yet (RX 6600 is not supported by HIP SDK and recompiling it on Windows is a pita), so I can either trust blindly or if you can show it working :) See review comments too.

blaisewf commented:
cool reviews, there is a huge difference in speed between zluda and dml?

AznamirWoW commented:
> Unfortunately I cannot test yet (RX 6600 is not supported by HIP SDK and recompiling it on Windows is a pita), so atm I can either trust blindly or if you can show it working :) See review comments too.

you just need a set of compiled libraries for gfx1032
https://cdn.discordapp.com/attachments/1207548125841989702/1262184101889441792/gfx1032.7z?ex=669f8f4c&is=669e3dcc&hm=522a86aee4e84f8294125ddd54ec7c631089269dcde7b7c25aaeea42a6b196e0&

deiteris commented:
> > Unfortunately I cannot test yet (RX 6600 is not supported by HIP SDK and recompiling it on Windows is a pita), so atm I can either trust blindly or if you can show it working :) See review comments too.
> 
> you just need a set of compiled libraries for gfx1032 https://cdn.discordapp.com/attachments/1207548125841989702/1262184101889441792/gfx1032.7z?ex=669f8f4c&is=669e3dcc&hm=522a86aee4e84f8294125ddd54ec7c631089269dcde7b7c25aaeea42a6b196e0&

Oh, thanks! I'm in the process of setting up everything, so I'll update shortly.

deiteris commented:
Doesn't seem to work correctly for me unfortunately. ZLUDA configuration seems to apply on startup:
![image](https://github.com/user-attachments/assets/b429ec49-f1d1-49c2-9868-fb620802b373)

And I can see ZLUDA device in Advanced Settings in training:
![image](https://github.com/user-attachments/assets/5cb94e34-7ad4-4abe-a8c7-51d13652eb1a)

But inference simply gets stuck. I see that GPU memory gets allocated, but no GPU usage occurs:
![image](https://github.com/user-attachments/assets/8e171699-4d82-4165-a9c3-77205a2ab0d8)

Here are the steps that I took:
1. Downloaded and installed HIP SDK from https://www.amd.com/en/developer/resources/rocm-hub/eula/licenses.html?filename=AMD-Software-PRO-Edition-24.Q3-Win10-Win11-For-HIP.exe.
2. Downloaded Tensil libraries that you provided and extracted them in `C:\Program Files\AMD\ROCm\6.1\bin\rocblas\library` (replaced files when prompted).
3. Downloaded and extracted ZLUDA from here: https://github.com/lshqqytiger/ZLUDA/releases/tag/rel.86cdab3b14b556e95eafe370b8e8a1a80e8d093b
4. Extracted ZLUDA in Applio folder for convenience.
    ![image](https://github.com/user-attachments/assets/6b9c02cd-02c3-41c0-b391-8120c40403dd)
5. Downloaded and installed Python 3.10.11.
6. Created venv with `python -m venv venv`.
7. Modified requirements to include torch built with CUDA 11.8.
    ![image](https://github.com/user-attachments/assets/c3a2ef08-5e1f-4053-ba50-8da1e12b2d58)
9. Installed requirements with `pip install -r requirements.txt`.
10. Copied ZLUDA DLLs with batch script in my venv folder:
      ```
      copy zluda\cublas.dll venv\Lib\site-packages\torch\lib\cublas64_11.dll /y
      copy zluda\cusparse.dll venv\Lib\site-packages\torch\lib\cusparse64_11.dll /y
      copy zluda\nvrtc.dll venv\Lib\site-packages\torch\lib\nvrtc64_112_0.dll /y
      ```
11. Finally, run with `.\zluda\zluda.exe -- python .\app.py`.

AznamirWoW commented:
> Doesn't seem to work correctly for me unfortunately. ZLUDA configuration seems to apply on startup: ![image](https://private-user-images.githubusercontent.com/6103913/351052815-b429ec49-f1d1-49c2-9868-fb620802b373.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NzE5ODQsIm5iZiI6MTcyMTY3MTY4NCwicGF0aCI6Ii82MTAzOTEzLzM1MTA1MjgxNS1iNDI5ZWM0OS1mMWQxLTQ5YzItOTg2OC1mYjYyMDgwMmIzNzMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjJUMTgwODA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmU3YWYxZTUxOTAzZGRiYjQxODdmZTEwNzI3OTViNDM2ZWQ0NWI2MzQzOTdkOWEyMjViMmI0NjdlODFiMWI0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.autNmIUEaKRznCDAP1W07y0TzU1jiVOlTdwfPiiF9Rs)
> 
> And I can see ZLUDA device in Advanced Settings in training: ![image](https://private-user-images.githubusercontent.com/6103913/351050843-5cb94e34-7ad4-4abe-a8c7-51d13652eb1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NzE5ODQsIm5iZiI6MTcyMTY3MTY4NCwicGF0aCI6Ii82MTAzOTEzLzM1MTA1MDg0My01Y2I5NGUzNC03YWQ0LTRhYmUtYThjNy01MWQxMzY1MmViMWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjJUMTgwODA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTE5NTA3M2JlNDA0ZTc0NmM1YWFhMmU2YjMwOTgxOGE0YTBkZjExY2ZjYjE5YWNhZTQwOTUyMjdiNjk4ZGMyOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.2VoORmX8mDzIClgjcyvDkEIpCNXozA9JgemMoq-Ks1k)
> 
> But inference simply gets stuck (I waited more than 1 min, but it just doesn't finish). Audio sample is around 5 seconds long, it should've been processed in no time (takes ~70ms with DirectML backend).
> 
> I see that GPU memory gets allocated, but no GPU usage occurs and there are no errors 😢: ![image](https://private-user-images.githubusercontent.com/6103913/351051047-8e171699-4d82-4165-a9c3-77205a2ab0d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NzE5ODQsIm5iZiI6MTcyMTY3MTY4NCwicGF0aCI6Ii82MTAzOTEzLzM1MTA1MTA0Ny04ZTE3MTY5OS00ZDgyLTQxNjUtYTljMy03NzIwNWEyYWIwZDgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjJUMTgwODA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTY4OWIyZTgwNTg2MzgzMWI2YWE0YTU5YzgzOWEyNDhhOTNmYjViYzU5ZTEzYzg5ZjFhMjM5MzYzYmI0NjM3OSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.48f_2SdTKyeN6S5LqhPt-whJtRW8Z47PcbtJu51UxSI)
> 
> Here are the steps that I took:
> 
>     1. Downloaded and installed HIP SDK from https://www.amd.com/en/developer/resources/rocm-hub/eula/licenses.html?filename=AMD-Software-PRO-Edition-24.Q3-Win10-Win11-For-HIP.exe.
> 
>     2. Downloaded Tensile libraries that you provided and extracted them in `C:\Program Files\AMD\ROCm\6.1\bin\rocblas\library` (replaced files when prompted).
> 
>     3. Downloaded ZLUDA from here: https://github.com/lshqqytiger/ZLUDA/releases/tag/rel.86cdab3b14b556e95eafe370b8e8a1a80e8d093b
> 
>     4. Extracted ZLUDA in Applio folder for convenience.
>        ![image](https://private-user-images.githubusercontent.com/6103913/351051599-6b9c02cd-02c3-41c0-b391-8120c40403dd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NzE5ODQsIm5iZiI6MTcyMTY3MTY4NCwicGF0aCI6Ii82MTAzOTEzLzM1MTA1MTU5OS02YjljMDJjZC0wMmMzLTQxYzAtYjM5MS04MTIwYzQwNDAzZGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjJUMTgwODA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDZjOGM1NTMwODcyMzNlOGU3ZmE0MzE1MTJkZjAwZTVjOTc4NDc5YWI5NDdiMWFjOGNhMjA4NWY3NWUzOTkwOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.lVvgptnPRGiURmjzYUQTj0iUwRyxXy26aiUgeUh6vOU)
> 
>     5. Downloaded and installed Python 3.10.11.
> 
>     6. Created venv with `python -m venv venv` and activated `.\venv\Scripts\Activate.ps1`.
> 
>     7. Modified requirements to include torch built with CUDA 11.8.
>        ![image](https://private-user-images.githubusercontent.com/6103913/351052235-c3a2ef08-5e1f-4053-ba50-8da1e12b2d58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NzE5ODQsIm5iZiI6MTcyMTY3MTY4NCwicGF0aCI6Ii82MTAzOTEzLzM1MTA1MjIzNS1jM2EyZWYwOC01ZTFmLTQwNTMtYmE1MC04ZGExZTEyYjJkNTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjJUMTgwODA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWMyZDk3NmVkMTRiYzI5YmQ3NjM1MGM1MjIyZDJhZDRkODRhMzhjZmJjOGM4NzQzZGE5MDkzYjM2YjlmZjllYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ByJV0N8X7IZIot9OWJpM_IjM-U_4h3ypPGjDxd1sTTU)
> 
>     8. Installed requirements with `pip install -r requirements.txt`.
> 
>     9. Copied ZLUDA DLLs with batch script in my venv folder:
>        ```
>        copy zluda\cublas.dll venv\Lib\site-packages\torch\lib\cublas64_11.dll /y
>        copy zluda\cusparse.dll venv\Lib\site-packages\torch\lib\cusparse64_11.dll /y
>        copy zluda\nvrtc.dll venv\Lib\site-packages\torch\lib\nvrtc64_112_0.dll /y
>        ```
> 
>     10. Finally, run with `.\zluda\zluda.exe -- python .\app.py`.

initial run for inference and training may take a while (5-20 minutes) because Zluda compiles the kernel code in the background.  That only happens once

AznamirWoW commented:
here's my benchmark

7600X CPU:

Converting audio 'assets\audios\12_min_audio.mp3'...
Conversion completed at 'assets\audios\12_min_audio_output.wav' in 433.26 seconds.

7600X CPU + 6700XT GPU / Zluda

Converting audio 'assets\audios\12_min_audio.mp3'...
Conversion completed at 'assets\audios\12_min_audio_output.wav' in 32.49 seconds.

deiteris commented:
Oh, you're right. It did take quite a while to compile (1226 seconds to be precise) and peaked at 4.2GB of RAM at some point. Inference seems to work well in general, at least I don't observe any issues with it.

Might be unrelated to ZLUDA, but inference consumes a lot of VRAM. For example, this is with 01:36 audio length. Takes around 12.8s to complete
![image](https://github.com/user-attachments/assets/5dfc13ac-d8d2-464f-843e-9c02f3055a4d)
```
Converting audio 'assets\audios\Pura_Pura_Lupa_VERSI_JEPANG_Mahen_-_Andi_Adinata_Cover.m4a_Vocals_Mel-RoFormer_No_Reverb.wav'...
Conversion completed at 'assets\audios\Pura_Pura_Lupa_VERSI_JEPANG_Mahen_-_Andi_Adinata_Cover.m4a_Vocals_Mel-RoFormer_No_Reverb_output.wav' in 11.57 seconds.
```

And this is with 04:21 audio length:
![image](https://github.com/user-attachments/assets/0374af27-d7ab-47ec-8823-49947afbf17d)
```
Converting audio 'assets\audios\Eminem_-_Lose_Yourself_vocals_only_explicit_content_MORE_intense_without_the_music__CHECK_IT.wav'...
Conversion completed at 'assets\audios\Eminem_-_Lose_Yourself_vocals_only_explicit_content_MORE_intense_without_the_music__CHECK_IT_output.wav' in 53.29 seconds.
```

AznamirWoW commented:
What's the point of having VRAM and not using it?

I'm more concerned about Applio holding ~4GB after inference is done

![image](https://github.com/user-attachments/assets/53679917-e4f4-40c9-91cb-d95dc1d1da9d)

deiteris commented:
> What's the point of having VRAM and not using it?

What I mean is that it has very negative performance impact. Once it hits VRAM limit, it goes for system RAM.

But anyway, someone with Nvidia GPU needs to confirm this behavior. This is out of scope of this PR.

blaisewf commented:
if all the changes are okay, we have to find an easy way for the users to run it

AznamirWoW commented:
> if all the changes are okay, we have to find an easy way for the users to run it

I have a .bat file that does the cu118 torch download and zluda patching

blaisewf commented:
> > if all the changes are okay, we have to find an easy way for the users to run it
> 
> I have a .bat file that does the cu118 torch download and zluda patching

we can add a run-applio-amd run-install-amd
